### PR TITLE
Integrate apply_optim_in_backward with DDP

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89194

Allow _apply_optim_in_backward to work with DDP.

Example:

```
dist.init_process_group("nccl", rank=rank, world_size=2)
    torch.cuda.set_device(rank)
    e = enc().cuda(rank)
    _apply_optimizer_in_backward(
        optimizer_class=torch.optim.SGD,
        params=e.parameters(),
        optimizer_kwargs={"lr": 0.03},
    )
    e = DDP(e, device_ids=[rank])
    inp = torch.randn(1, 10, device=rank)
    e(inp).sum().backward()
```

Constraints:

1. Custom communication hook not yet supported
2. _apply_optim_in_backward needs to be called _before_ wrapping model in DDP.
3. DDP will remove the gradient hooks _apply_optim_in_backward registers, so these gradient hooks will not be fired and cannot be used.
4. All DDP managed parameters have grads set to None by default once optimizer is applied. There is no support for setting only some parameter grads to None, this must be done manually by user (and DDP_OVERLAPPED_OPTIM_SET_GRADS_TO_NONE=0 needs to be set.)

Differential Revision: [D41329694](https://our.internmc.facebook.com/intern/diff/D41329694/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41329694/)!

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -690,6 +696,41 @@ def __init__(
...
 
+        # Check if user has used apply_optim_in_backward to overlap optimizer
+        # step + DDP backward. Currnent constraints:
+        # 1. Only allreduce is supported at the moment, no custom communication.
+        # 2. The reducer by default sets all grads for parameters DDP manages to
+        # None after they have been applied by the optimizer. There is no support
+        # for setting only some parameter grads to None, this must be done manually
+        # by user (and DDP_OVERLAPPED_OPTIM_SET_GRADS_TO_NONE=0 needs to be set.)
+        if any(
+            hasattr(p, '_in_backward_optimizers') for p in self.module.parameters()
```
Comment (awgu) [inline torch/nn/parallel/distributed.py:707]: Should we filter ignored parameters instead of directly using `self.module.parameters()`?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -690,6 +696,41 @@ def __init__(
...
+        ):
+            # Remove hooks that apply_optim_in_backward had registered because
+            # DDP customizes how optimizer is overlapped with backward due to
+            # the allreduce.
+            for p in self.module.parameters():
```
Comment (awgu) [inline torch/nn/parallel/distributed.py:712]: ditto about ignored parameters

File: torch/testing/_internal/distributed/distributed_test.py [hunk 3]
```diff
@@ -31,6 +31,7 @@
     default_hooks as default,
     quantization as quantization_hooks,
 )
+from torch.distributed.optim import _apply_optimizer_in_backward
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:34]: Do these added unit tests cover both `DDP_OVERLAPPED_OPTIM_SET_GRADS_TO_NONE = 1` and `0` cases? I am not seeing the `0` case.

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -459,6 +459,9 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
             return reducer.get_grad_buckets(/* return_zero_tensors */ true);
           },
           py::call_guard<py::gil_scoped_release>())
+      .def("_set_grads_to_none", [](::c10d::Reducer& reducer) {
+          reducer.set_grads_to_none(true);
```
Comment (awgu) [inline torch/csrc/distributed/c10d/init.cpp:463]: Conceptual: What does the `true` in `reduce.set_grads_to_none(true)` mean?

File: torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py [hunk 5]
```diff
@@ -32,6 +32,32 @@ def _check_valid_functional_optim(self):
             )
 
 
+def _apply_optim_in_backward_hook(
```
Comment (awgu) [inline torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:35]: Should we rename this to suggest that it is a factory for the hook?

Otherwise, the only difference between this method and the actual hook defined inside is the preceding `_`, which may be confusing since the two have different quite semantics (one being a factory and one being the hook to run itself).

File: torch/testing/_internal/distributed/distributed_test.py [hunk 6]
```diff
@@ -4562,6 +4563,67 @@ def test_ddp_hook_with_optimizer_parity_sgd(self, optimize_subset):
                 weight_decay=sgd_weight_decay,
             )
 
+        def _test_ddp_apply_optim_in_backward(
+            self, optim_cls, optim_kwargs, register_hook=True,
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:4567]: Why make `register_hook` a default argument?

We could just not default it, and then explicitly pass `True` or `False` e.g. as part of a subtest.

**Review by awgu — COMMENTED** [review]
Could we explain the thought process that concluded to use an environment variable `DDP_OVERLAPPED_OPTIM_SET_GRADS_TO_NONE` to configure behavior?
- The comment in the code suggests that it is due to the experimental nature, but I am wondering what is the tradeoff compared to a private argument.
- Due to my lack of experience and not really having seen this before, I am wondering if using an environment variable like this has precedent and is a reasonable practice.

Also, for these WIP PRs, it would be helpful if you could give a pointer to what we should be focusing on when we review. For example, are there particular design decisions that you want to discuss? Or is it ready to be fully nitpicked for code quality?

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -697,22 +703,24 @@ def __init__(
...
         if any(
             hasattr(p, '_in_backward_optimizers') for p in self.module.parameters()
         ):
             # Remove hooks that apply_optim_in_backward had registered because
             # DDP customizes how optimizer is overlapped with backward due to
             # the allreduce.
             for p in self.module.parameters():
-                for handle in p._optimizer_hook_handles:
+                for handle in getattr(p, '_optimizer_hook_handles', []):
                     handle.remove()
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:714]: if we remove the handle, why we require users to pass register_hook=False then? could we enforce users to call apply_optimizer_in_backward() before DDP wrapping? will that resolve the issue without adding the 'register_hook' flag?

File: torch/csrc/distributed/c10d/init.cpp (hunk 4 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/init.cpp:463]: It means that we should set the grads to None.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:707]: yes, great catch!

**Comment by rohan-varma** [comment]
@awgu For DDP_OVERLAPPED_OPTIM_SET_GRADS_TO_NONE, it is mostly a killswitch that can quickly disable setting grads to None for DDP or not.

The default behavior is that the reducer will set _all_ DDP grads to None when running with overlapped optimizer. This is of course an issue if user might want to use non-overlapped optimizer for some DDP managed parameters (why user may want to do some overlapped optimizer and some not is unclear to me though). 

In the future, we could consider proper support where DDP sets only the explicit parameters which have in backwards optimizers to None, but we haven't seen a use case requiring this yet. I filed an issue for future support: https://github.com/pytorch/pytorch/issues/90052

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py]: Discussed this offline, this reasoning makes sense and is the direction we want to converge on, but some internal workloads are running into issues we need to further debug.

For now, we will go with this approach and follow up to ultimately consolidate on either doing `register_hook=False`, or removing the handles.

**Review by zhaojuanmao — APPROVED** [review]
looks good to me! thank you

**Comment by zhaojuanmao** [comment]
some lint error and unit test failed

**Comment by rohan-varma** [comment]
lint jobs are unrelated and failing due to clang-tidy binary hash issue.
